Start linting the examples folder

MVrachev opened the pull request:
Fixes #1697

**Description of the changes being introduced by the pull request**:

The examples folder currently contains a repository example and it's
good if we start linting its content and as a result add type
annotations.

I run all of the linters on the example folder instead of only using `mypy`
as I thought that will be better for the code included there.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Start linting the examples folder

The examples folder currently contains a repository example and it's
good if we start linting its content and as a result add type
annotations.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

lukpueh commented on examples/repo_example/basic_repo.py:
```diff
@@ -332,7 +333,7 @@ def _in(days):
 
 # Increase expiry (delegators should be less volatile)
-roles["targets"].expires = _in(365)
+roles["targets"].signed.expires = _in(365)
```
👀 Whoopsi, thanks for catching this! Did mypy detect it?

MVrachev replied:
Yes, it did.

lukpueh commented on examples/repo_example/basic_repo.py:
```diff
@@ -117,7 +118,7 @@ def _in(days):
 target_path = f"{local_path.parts[-2]}/{local_path.parts[-1]}"
 
-target_file_info = TargetFile.from_file(target_path, local_path)
+target_file_info = TargetFile.from_file(target_path, str(local_path))
```
Was this also flagged by mypy?

(Also it is odd, because I remember already casting this to `str` at some point because `TargetFile.from_file` raised an error with local_path being a `Path` object. But I can't find it in #1685. I wonder were it went... Anyways, good that you/mypy fixed it.)

MVrachev replied:
Yes, `mypy` noticed that.
Yet another argument as to why we do need linting quickly as possible on new code.

jku commented on pyproject.toml:
```diff
@@ -79,7 +79,8 @@ show_error_codes = "True"
-  "tuf/exceptions.py"
+  "tuf/exceptions.py",
+  "examples/",
 ]
```
left a comment about this in the other bug: https://github.com/theupdateframework/python-tuf/pull/1699#discussion_r760934677 (we could remove these and just define the "linter dirs" in one place in tox.ini to make maintenance easier)

lukpueh commented on examples/repo_example/basic_repo.py:
```diff
@@ -89,8 +90,8 @@ def _in(days):
-roles = {}
-keys = {}
+roles: Dict[str, Metadata] = {}
+keys: Dict[str, Dict[str, Any]] = {}
```
Quick question, what is the criterion for a variable name to get type annotated?

jku replied:
it's needed if mypy can't figure it out from the context (and mypy is not very advanced in this). So in this case `roles = { "root": root_md }` would have been fine but an empty container needs typing

MVrachev replied:
I think it does need annotation when you use containers and personally, I think it's better to define annotations in these cases. It makes it so much easier to understand how the container should be used or is used later.

lukpueh reviewed (commented):
Thanks for the PR, @MVrachev! From a quick glance it looks good to me! I'll take another look when it no longer has the draft status, which I think you can remove now that #1699 has been merged.

Commit:
Linting in tox.ini: use variable for target dirs

Instead of providing a target directory for linting by each of the
tools use one variable which will be the source of truth about which
directories do we lint.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
I rebased on top of develop and created a commit adding a `lint_dirs` variable as proposed by Jussi.
@lukpueh is right that this list will be simplified after we drop the old code base, but still, it will be useful to have
a `lint_dirs` variable even if the list is simply `./`

This pr is ready for review.

lukpueh reviewed (approved):
Works like a charm!
